### DISCUSSION TITLE
Qt6 static type checking issues, decisions, etc

### DISCUSSION THREAD

**Discussion by toofar** [qutebrowser/qutebrowser#7372] [General]
Threads relating to https://github.com/qutebrowser/qutebrowser/issues/7370

For things like:

* how best to solve this scenario that mypy is complaining about
* how best to solve a conflict between PyQt5 and PyQt6 runs
* how can we make this use case easier going forward (eg new contributes generally aren't expected to run all the permutations, should we set a default?)
* making decisions (eg if VS Code decides to use PySide for autocompletion is that good enough?)

*Triaging issues mypy raises*

There are quite a lot of them. It probably makes sense to do all the _easy_ ones first to cut down the number and then discuss the tricky ones here.

When solving issues it'll probably help speed things up to have two terminals open with, one running over PyQt6 and one over PyQt5. So that you can confirm you aren't breaking the other one.
Possibly the option to include/exclude certain files would make working though a single file easier.
Maybe the mypy daemon can make stuff even faster.

**Comment by toofar** [comment]
Should we set a default binding/wrapper for mypy? If so why should it be PyQt6?

This would be so just running `mypy qutebrowser` isn't basically completely useless.

Eg, in .mypy.ini
```ini
[mypy]
always_true = USE_PYQT6,IS_QT6
always_false = USE_PYQT5,USE_PYSIDE2,USE_PYSIDE6,IS_QT
```

We will in any case be setting up tox targets like `tox -e mypy-qt6` and `tox -e mypy-qt5`

**Reply by toofar** [reply]
Nevermind. This doesn't seem to work:

    mypy: error: You can't make a variable always true and always false (IS_QT5, IS_QT6, USE_PYQT5, USE_PYQT6)
    
Maybe we could ask about this use case on the mypy issue tracker.

Well we could probably set a default in the file and remove it with sed before running in CI but that's a bit yuck.

**Reply by The-Compiler** [reply]
Hmm yeah, bummer that doesn't work. Looks like they handle overriding those things on a case-by-case basis, see e.g. https://github.com/python/mypy/commit/2ac722ac7d844f4c5ae2f3c36f3470543cf01b11.

I don't think we necessarily need to support running mypy directly (though I suppose some IDEs and such still attempt that nowadays?). Maybe it's okay if the only supported interface to run it is via tox, in which case we might be able to get a bit more creative. Though I'm not sure how off-hand. Worst-case I suppose we can always run a wrapper script (if we decide that via tox is the only supported invocation anyways, it really doesn't matter?).

I suppose the question we might want to answer first is whether we do want to type check with multiple backends. Over in the issue you seem to imply yes, but I'm honestly not sure - I'll start another sub-thread on that, though.

If we decide we do not, having an `if TYPE_CHECKING:` as the first conditional might be enough.

**Comment by toofar** [comment]
How to get autocompletion working with VS Code?
Edit: resolved :heavy_check_mark: 

At this point I would recommend reversing the order of the conditionals in the files like qutebrowser/qt/webenginecore.py as that seems to avoid the issue for whatever reason (more notes in the actual issue). The actual preference is done in `machinery._autoselect_wrapper()` anyhow. (And maybe removing PySide2 support.)

Any other environments we should make a point to support? If anyone uses Deoplete with (neo)vim feel free to see if it works for you there? I've been checking with QWebEngineHistory in qutebrowser/qt/webenginecore.py as that's one of the affected objects.

We should still raise an issue with pyright asking if the can make `defineConstant` handle this case.

**Reply by The-Compiler** [reply]
Agreed on switching the order - I really don't mind much in which order things are, and if it makes things better, let's go for it.

Also agreed on dropping PySide2 support. Really not worth bothering with it at this point. Would like to see PySide6 in the future, but I guess nobody will care about PySide2 (vs. just using PyQt5) enough for it to be worth the trouble, especially since various functionality will be broken with how many APIs are missing in it.

@toofar do you plan to work on is, or should I take a look?

**Reply by toofar** [reply]
I don't think I'm going to do any real work this week (might tidy up the autocompletion part of the linked issue but it's spring here on Thursday and my garden needs some attention).
Also, assuming its just dropping a few lines and re-ordering stuff, I kinda want you to work on it so you can make sure it actually makes thing work better for you.

**Reply by toofar** [reply]
Pushed to the qt6-v2 branch https://github.com/qutebrowser/qutebrowser/commit/6f8d7a7fbc6350b16046b832e8f5957b349245f2 and https://github.com/qutebrowser/qutebrowser/commit/4cf6aa8cef5227d3610b31144318cd78109d2809
Tested with VS Code in flatpack. Pretty sure it was back at all default settings.

**Reply by toofar** [reply]
Looks like pylance has been updated now. I pushed a commit that adds a pyrightconfig.json file tuned for PyQt5: https://github.com/qutebrowser/qutebrowser/commit/97920c8271170b1347b6b3cd351790959589ab6a

**Comment by The-Compiler** [comment]
Do we really want to have type checking for multiple backends?

@toofar you seem to imply yes in the issue, but I'm not entirely convinced:

- The official PyQt stubs [have various issues](https://www.riverbankcomputing.com/pipermail/pyqt/2022-August/044840.html) currently
- The [unofficial PyQt5 stubs](https://github.com/python-qt-tools/PyQt5-stubs/) work way better - but even there, sometimes we need to have type ignores because the stubs just aren't there yet.
- There will be [unofficial PyQt6 stubs](https://github.com/python-qt-tools/PyQt6-stubs) (and there are discussions to just include those fixes in PyQt6 proper starting, thankfully - though they seem to have [died down again](https://www.riverbankcomputing.com/pipermail/pyqt/2022-August/044850.html))
- We will likely get into situations where we need to ignore an error for one kind of stubs but not the other. Or the other way around. Does mypy give us any facilities to do that (I don't think so)? How do we handle this, especially since we have `warn_unused_ignores` turned on (and I'd kinda like to keep that on)?

All in all, I feel like this will be a real uphill battle... I wonder if I'd be better to fix the PyQt5 mypy run, and then maybe only take a manual look at the PyQt6 issues. If we decide they are indeed fixable without breaking PyQt5 again, maybe we can indeed run both and I'm wrong.

Then at some point in the future, we can switch over to PyQt6 once there are actually usable stubs for it released somewhere maybe? I guess PyQt5 support will be pretty much "on life support" at some point anyways - note that even the LTS support [will end in mid 2023](https://wiki.qt.io/Qt_5.15_Release). Though I don't know how much longer [KDE will maintain it](https://community.kde.org/Qt5PatchCollection), and it will take another year until the final 5.15 release is actually opensource thanks to Qt's dumb new LTS policies...

I can't really say at this point when we'll drop Qt 5 support, which will of course make the decision much easier :wink:. I'd definitely like to keep it around for 3.0.0 though, I think.

**Reply by toofar** [reply]
I would like to (and did!) present type checking both backends as the strawman proposal. Because I think it would be good to be checking everything we could. I don't know how hard it would be. Some thoughts:

I guess I'm more optimistic about the state of the PyQt6 type hinting, both the unofficial stubs and the riverbank stuff. I've been using [my fork](https://github.com/toofar/PyQt6-stubs/commits/main) of the unofficial stubs (which just naively adds webengine and upgrades the PyQt version) and, while I haven't looked though the issues mypy raises in much detail, skimming through them they all look like they could be correctly raised issues (the majority of them are relating to webkit, which I don't know how to resolve, maybe exclude whole files...). I haven't compared then to just the types included in the PyQt6 release, but Phil sounded pretty confident when he said they were good!

I guess I would like to take an approach that hedges our bets. Eg support one then look at both afterwards. How that looks depends on which one you think should be the default.

* Get PyQt5 green first -> Do that, create a qt6-mypy branch, try to get that working without breaking the PyQt5 type hints, if that seems to hard then get it green and switch over in one go (I guess that is what you said anyway in your summary!)
* Get PyQt6 working first -> Whether we even bother with type checking PyQt5 depends on how long it'll live. Going on your uncertainty above it seems like it might be worth trying to get it working if we go this route too (look how long we've asked people to consider webkit when making changes!), in this case too have a crack a both but give up if its too hard

Regarding ignores needed with one and unnecessary on the other one... yeah. I haven't looked into any strategies to handle that yet. I've ran into it already with [these ones](https://github.com/qutebrowser/qutebrowser/blob/0877fb0/qutebrowser/utils/utils.py#L141-L151). Having the different implementation per backend (eg `if PYQT6: dothing-qt6(...) elif PYQT5: dothing-qt5(...)`) is an option but I suspect this will happen too often for that to be palatable.

So I'm fine with focusing on getting PyQt5 type checking going again and looking at PyQt6 later (and switching to that alone if we can't get them both working without a lot of pain).

**Reply by toofar** [reply]
PR for pyqt5 here: https://github.com/qutebrowser/qutebrowser/pull/7392
Remains to be seen whether we can get PyQt6 working in the same codebase or whether it'll be on a branch for a while then a big switchover.

**Comment by The-Compiler** [comment]
How much does code churn really matter?

@toofar you write:

> we can attempt re-writing tools to help people rebase/merge their but the edge cases are huge paint, nothing really complete out of the box that can help us here atm

Can you elaborate a bit on the state of #7128 (or the v2) regarding that? I like the solution for its simplicity as far as the code goes. I dislike the verbosity of changed code.

I'm a bit undecided as for the code churn. It sucks, yes - but also we have code churn because of the enums already anyways, and we plan to run black at some point... so if we had a good autoformatting thing (even if it just gets 90% there) we can run on every PR somehow, then maybe that will just be a necessary evil?

But all things considered, I believe I still prefer the current approach, if we can make it work across setups without too much pain for devs.

**Reply by toofar** [reply]
>> the edge cases are a huge pain, nothing really complete out of the box that can help us here atm

> Can you elaborate a bit on the state of #7128 (or the v2) regarding that?
> if we had a good autoformatting thing (even if it just gets 90% there) we can run on every PR somehow, then maybe that will just be a necessary evil?

I've got bad memories have having to re-apply manual fixes on top of the changes the rename script made while rebasing (and basically redoing everything) a couple of times. But now you've prompted me to have another look it seems it is idempotent if you [hold it
right](https://github.com/qutebrowser/qutebrowser/commit/a7cdd23c005ae2673c8b066c60bd1d714db749e6).

So I guess most people would be unlikely to be introducing new edge cases on their branches (mostly around conditional imports and imports halfway down the file and imports for side effects without using the imported thing).

It still takes 15 minutes to run on my desktop though. 

Apart from the speed, how would they run it anyway? If they ran it on their branch before merging then it would run into all those edge cases we manually fixed when they merged later. If you blindly merge with an "ours" strategy or something and then run it later you are probably going to also introduce a bunch of issues with normal drift. Possibly doing a rebase with "exec rename" applied to each commit on the branch would be cleanest but doing a rebase when you have got a year or so of drift can really suck your enthusiasm. idk, easier to avoid the problem I think.

The script is [here](https://github.com/qutebrowser/qutebrowser/blob/feat/pyqt_import_rewrite_v2/scripts/codemods/rename_pyqt.py) and it's unchanged from the first branch. I just did a new one because you had changed all the imports (from PyQt5 to qutebrowser.qt) already so a rebase would have been a pain.

> I like the solution for its simplicity as far as the code goes.

Which part of it? Do you mean the style of importing (`from qute.qt import
core; core.pyqtSignal(...)`) or that [qt.py](https://github.com/qutebrowser/qutebrowser/blob/feat/pyqt_import_rewrite_v2/qutebrowser/qt.py) is a file with named imports instead of modules with wildcard imports?

I'll attempt some pros/cons of how I see it:

**qt.py as file**

| pros | cons |
| --- | --- |
| it's a single file? | less flexible importing, can't do `from qute.qt.core import pyqtSignal` |
|    | I don't think mypy handles the moved webengine modules how I've done them (delattr, setattr on the modules) |
 
And a fair amount of stuff that relies on ImportErrors to gate behaviour stops working and has to be changed to explicitly check that stuff isn't None.

**qt as a package**

| pros | cons |
| --- | --- |
| import stuff any way you like | wildcard imports are slow (more stuff is now being imported) and can be hard to follow |

> But all things considered, I believe I still prefer the current approach, if we can make it work across setups without too much pain for devs.

+1
